test-data/unit/check-typeddict.test

```diff
@@ -863,6 +863,28 @@ c: Union[A, B] = {'@type': 'a-type', 'a': 'Test'}
 reveal_type(c) # N: Revealed type is "Union[TypedDict('__main__.A', {'@type': Literal['a-type'], 'a': builtins.str}), TypedDict('__main__.B', {'@type': Literal['b-type'], 'b': builtins.int})]"
 [builtins fixtures/tuple.pyi]
 
+[case testTypedDictUnionUnambiguousCaseBasedOnType]
+from typing import Union, Mapping, Any, cast
+from typing_extensions import TypedDict, Literal
+
+A = TypedDict('A', {'@type': Literal['a-type']})
+B = TypedDict('B', {'@type': Literal['b-type']})
+C1 = TypedDict('C1', {'@type': Literal['c-type'], 'extra': int})
+C2 = TypedDict('C2', {'@type': Literal['c-type'], 'extra': str})
+
+x: Union[A, B, C1, C2]
+x = {'@type': 'a-type'}
```

Comment:
Maybe I'm missing something, but in the test case above `[case testTypedDictUnionUnambiguousCase]` - the inferred type of `c: Union[A, B] = {'@type': 'a-type', 'a': 'Test'}` is `Union[TypedDict('__main__.A', {'@type': Literal['a-type'], 'a': builtins.str}), TypedDict('__main__.B', {'@type': Literal['b-type'], 'b': builtins.int})]` which looks inconsistent with this case. 

Should it be this way?

Reply:
There is a difference if it's annotated directly `c: Union[A, B] = ...` or separately `c: Union[A, B]; c = ...`. Initially I emulated the style of the previous test (annotated directly), but then switched to this one because it's shorter. I cannot explain why the difference exist, naively I would have thought they should be equivalent. I can try to figure it out if you think it's pertinent.

Reply:
> naively I would have thought they should be equivalent

Yeah, me too! 😄  I would be very confused if I saw this. I don't remember any other cases when `x: Annotation; x = ...` and `x: Annotation = ...` would produce different results.

Reply:
For reference, this behavior is not particular to TypedDict:

```py
class A: pass
class B: pass

x: A|B
x = A()
reveal_type(x)  # Revealed type is A

y: A|B = A()
reveal_type(y)  # Revealed type is Union[A, B]
```

Reply:
Yeah this behavior is suprising. We have it for historical reasons and because without this feature there would be no easy way to force a wider/more general type for a variable in a type-safe manner.